src/codegen/compile.ml

```diff
@@ -10188,6 +10190,19 @@ and compile_dec env pre_ae how v2en dec : VarEnv.t * G.t * (VarEnv.t -> scope_wr
   (fun (pre_ae, alloc_code, mk_code, wrap) ->
        G.(pre_ae, with_region dec.at alloc_code, fun ae body_code ->
           with_region dec.at (mk_code ae) ^^ wrap body_code)) @@
+
+  (* Set up some helpers, for exclusive usage by the "constant expressions" special case below *)
+  let const_exp_helper, irrefutable_helper =
+    lazy begin
+        let[@warning "-8"] LetD (p, e) = dec.it in
+        const_exp_matches_pat env p e
+      end,
+    lazy begin
+        let[@warning "-8"] LetD (p, _) = dec.it in
+        Ir_utils.is_irrefutable p
+      end in
+  let compile_time_matchable pred = Lazy.(force irrefutable_helper || pred (force const_exp_helper)) in
+
```

Comment:
This looks awfully complicated to me. Isn't irrefutably just a special case of a sure-to-pass match?

So maybe have a single function for matching a pattern against a const value that returns the ternary result, and no longer treat irrefutably separately?

Reply:
Also the higher order pred seems a bit roundabout. Maybe introduce a data type for will_match, will_fail or unknown, call it once, and then match on the result?

Reply:
PTAL!